### PR TITLE
Provide logical messages for authentication errors on API 

### PR DESCRIPTION
**Pull request by louis878**
**Is your feature request related to a problem? Please describe.**
When a user provides a faulty authentication/authorization token he should receive a logical error message. This can be done by capturing the JWT Runtime Exceptions in a ControllerAdvice

### CODE REVIEW

File: core/java/iesi-rest-without-microservices/src/main/java/io/metadew/iesi/server/rest/configuration/security/jwt/ApiError.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import lombok.Setter;
+import org.springframework.http.HttpStatus;
+
+import java.time.LocalDateTime;
+
+@Getter
+@Setter
+public class ApiError {
+
+    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "dd-MM-yyyy hh:mm:ss")
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-rest-without-microservices/src/main/java/io/metadew/iesi/server/rest/configuration/security/jwt/ApiError.java:18]: Does this line do the same as line 32-33?

File: core/java/iesi-rest-without-microservices/src/main/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilter.java [hunk 2]
```diff
@@ -23,14 +25,26 @@ public JWTAuthenticationFilter(JWTAuthenticationConverter jwtAuthenticationConve
     }
 
     @Override
-    protected void doFilterInternal(HttpServletRequest httpServletRequest, HttpServletResponse httpServletResponse, FilterChain filterChain) throws ServletException, IOException {
-        Authentication authentication = jwtAuthenticationConverter.convert(httpServletRequest);
-        if (authentication == null) {
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-rest-without-microservices/src/main/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilter.java:28]: Is this line not needed (anymore). Do we need to check if it is possible that a null Authentication is returned by jwtAuthenticationConverter.convert?

File: core/java/iesi-rest-without-microservices/src/main/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilter.java [hunk 3]
```diff
@@ -23,14 +25,26 @@ public JWTAuthenticationFilter(JWTAuthenticationConverter jwtAuthenticationConve
...
+    public void doFilterInternal(HttpServletRequest httpServletRequest, HttpServletResponse httpServletResponse, FilterChain filterChain) throws IOException, ServletException {
+        try {
+            Authentication authentication = jwtAuthenticationConverter.convert(httpServletRequest);
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-rest-without-microservices/src/main/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilter.java:30]: only surround the jwtAuthenticationConverter.convert(httpServletRequest) with a try-catch clause

File: core/java/iesi-rest-without-microservices/src/main/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilter.java [hunk 4]
```diff
@@ -23,14 +25,26 @@ public JWTAuthenticationFilter(JWTAuthenticationConverter jwtAuthenticationConve
...
+    public void setErrorResponse(HttpStatus status, HttpServletResponse response, Throwable ex) {
+        response.setStatus(status.value());
+        response.setContentType("application/json");
+        ApiError apiError = new ApiError(status, ex);
+        try {
+            String json = apiError.convertToJson();
+            response.getWriter().write(json);
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-rest-without-microservices/src/main/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilter.java:48]: let the method setErrorResponse throw the IOException instead of just printing the stackTrace

File: core/java/iesi-rest-without-microservices/src/main/java/io/metadew/iesi/server/rest/configuration/security/jwt/JwtService.java [hunk 5]
```diff
@@ -32,14 +33,17 @@
...
-        return verifier.verify(token);
+        try {
+            Algorithm algorithm = Algorithm.HMAC256(secret);
+            JWTVerifier verifier = JWT.require(algorithm)
+                    .withIssuer(ISSUER)
+                    .build();
+            return verifier.verify(token);
+        } catch (JWTVerificationException e) {
+            throw new JWTVerificationException(e.toString());
+        }
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-rest-without-microservices/src/main/java/io/metadew/iesi/server/rest/configuration/security/jwt/JwtService.java:44]: Why is this needed?

File: core/java/iesi-rest-without-microservices/src/main/java/io/metadew/iesi/server/rest/configuration/security/jwt/JwtWebSecurityConfiguration.java [hunk 6]
```diff
@@ -74,6 +74,7 @@ protected void configure(HttpSecurity http) throws Exception {
                 .csrf().disable()
                 .authorizeRequests()
                 .mvcMatchers("/users/login").permitAll()
+                .mvcMatchers("/users/create").permitAll()
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-rest-without-microservices/src/main/java/io/metadew/iesi/server/rest/configuration/security/jwt/JwtWebSecurityConfiguration.java:77]: should not be permitted for now

File: core/java/iesi-rest-without-microservices/src/test/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilterTest.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        String token = "eyJ0eXAiOJKV1QiLCUzI1NiJ9.eyJzdWIiOiJMyIiwiaXNzIjoiaWVzaSIsImV4cCI6MTU5ODU2ODM2MywiaWF0IjoxNTk4NDY4MzYzLCJhdXRob3JpdGllcyI6W119.alYmNwRyiqS8VijXdQ2F0ooVdr5KkITz2bycffG6_fI";
+        MockHttpServletRequest request = new MockHttpServletRequest("GET", "/scripts");
+        request.addHeader(HttpHeaders.AUTHORIZATION, "Bearer " + token);
+        JWTAuthenticationFilter jwtAuthenticationFilter = new JWTAuthenticationFilter(jwtAuthenticationConverter);
+        MockHttpServletResponse response = new MockHttpServletResponse();
+        MockFilterChain chain = new MockFilterChain();
+
+        jwtAuthenticationFilter.doFilterInternal(request, response, chain);
+        assertThat(response.getStatus()).isEqualTo(401);
+        assertThat(response.getContentAsString()).contains("com.auth0.jwt.exceptions.JWTDecodeException: The string");
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-rest-without-microservices/src/test/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilterTest.java:36]: is it possible to do a jsonPath assertion on the json string? As can you check that the 'message' field starts with "com.auth0..." and the 'status' field is equal to 401.

File: core/java/iesi-rest-without-microservices/src/test/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilterTest.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        String token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE0Nzc1OTJ9.isvT0Pqx0yjnZk53mUFSeYFJLDs-Ls9IsNAm86gIdZo";
+        MockHttpServletRequest request = new MockHttpServletRequest("GET", "/scripts");
+        request.addHeader(HttpHeaders.AUTHORIZATION, "Bearer " + token);
+        JWTAuthenticationFilter jwtAuthenticationFilter = new JWTAuthenticationFilter(jwtAuthenticationConverter);
+        MockHttpServletResponse response = new MockHttpServletResponse();
+        MockFilterChain chain = new MockFilterChain();
+
+        jwtAuthenticationFilter.doFilterInternal(request, response, chain);
+        assertThat(response.getStatus()).isEqualTo(401);
+        assertThat(response.getContentAsString()).contains("com.auth0.jwt.exceptions.InvalidClaimException: The Claim");
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-rest-without-microservices/src/test/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilterTest.java:51]: is it possible to do a jsonPath assertion on the json string? As can you check that the 'message' field starts with "com.auth0..." and the 'status' field is equal to 401.

File: core/java/iesi-rest-without-microservices/src/test/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilterTest.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        String token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXUyJ9.eyJpc3MiOiJhdXRoMCJ9.AbIJTDMFc75MhvcP03nJPyCPzZtQcGEp-zWfOkEE";
+        MockHttpServletRequest request = new MockHttpServletRequest("GET", "/scripts");
+        request.addHeader(HttpHeaders.AUTHORIZATION, "Bearer " + token);
+        JWTAuthenticationFilter jwtAuthenticationFilter = new JWTAuthenticationFilter(jwtAuthenticationConverter);
+        MockHttpServletResponse response = new MockHttpServletResponse();
+        MockFilterChain chain = new MockFilterChain();
+
+        jwtAuthenticationFilter.doFilterInternal(request, response, chain);
+        assertThat(response.getStatus()).isEqualTo(401);
+        assertThat(response.getContentAsString()).contains("com.auth0.jwt.exceptions.SignatureVerificationException: The Token's Signature resulted invalid when verified using the Algorithm: HmacSHA256");
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-rest-without-microservices/src/test/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilterTest.java:66]: is it possible to do a jsonPath assertion on the json string? As can you check that the 'message' field starts with "com.auth0..." and the 'status' field is equal to 401.

File: core/java/iesi-rest-without-microservices/src/test/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilterTest.java [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+        String token = "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJpZXNpIiwiZXhwIjoxNTk4NjA2NTY4fQ.F7GnAfcklY1-0sG9K4tkz3mfVQ40aAtBG7MUdSI79FI";
+        MockHttpServletRequest request = new MockHttpServletRequest("GET", "/scripts");
+        request.addHeader(HttpHeaders.AUTHORIZATION, "Bearer " + token);
+        JWTAuthenticationFilter jwtAuthenticationFilter = new JWTAuthenticationFilter(jwtAuthenticationConverter);
+        MockHttpServletResponse response = new MockHttpServletResponse();
+        MockFilterChain chain = new MockFilterChain();
+
+        jwtAuthenticationFilter.doFilterInternal(request, response, chain);
+        assertThat(response.getStatus()).isEqualTo(401);
+        assertThat(response.getContentAsString()).contains("com.auth0.jwt.exceptions.TokenExpiredException: The Token has expired ");
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-rest-without-microservices/src/test/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilterTest.java:80]: is it possible to do a jsonPath assertion on the json string? As can you check that the 'message' field starts with "com.auth0..." and the 'status' field is equal to 401.

File: core/java/iesi-rest-without-microservices/src/main/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilter.java (hunk 2 above)
Comment (louis878) [reply core/java/iesi-rest-without-microservices/src/main/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilter.java:28]: Wil give a 403 in any case

File: core/java/iesi-rest-without-microservices/src/main/java/io/metadew/iesi/server/rest/configuration/security/jwt/ApiError.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+    public String convertToJson() throws IOException {
+        try {
+            ObjectMapper mapper = new ObjectMapper();
+            mapper.registerModule(new JavaTimeModule());
+            mapper.disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);
+            return mapper.writeValueAsString(this);
+        } catch (IOException e) {
+            e.printStackTrace();
+            throw new IOException(e);
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-rest-without-microservices/src/main/java/io/metadew/iesi/server/rest/configuration/security/jwt/ApiError.java:37]: you throw the Exception in the method definition and catch in the method execution. Please delete the capture in the method execution

File: core/java/iesi-rest-without-microservices/src/test/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilterTest.java [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+        MockHttpServletRequest request = new MockHttpServletRequest("GET", "/scripts");
+        request.addHeader(HttpHeaders.AUTHORIZATION, "Bearer " + token);
+        JWTAuthenticationFilter jwtAuthenticationFilter = new JWTAuthenticationFilter(jwtAuthenticationConverter);
+        MockHttpServletResponse response = new MockHttpServletResponse();
+        MockFilterChain chain = new MockFilterChain();
+
+        jwtAuthenticationFilter.doFilterInternal(request, response, chain);
+        assertThat(response.getStatus()).isEqualTo(401);
+        String json = JsonPath.read(response.getContentAsString(), "message");
+        assertThat(json).startsWith("The Token's Signature resulted invalid when verified using the Algorithm: HmacSHA256");
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-rest-without-microservices/src/test/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilterTest.java:70]: also check the status part of the message returned

File: core/java/iesi-rest-without-microservices/src/test/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilterTest.java [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+        request.addHeader(HttpHeaders.AUTHORIZATION, "Bearer " + token);
+        JWTAuthenticationFilter jwtAuthenticationFilter = new JWTAuthenticationFilter(jwtAuthenticationConverter);
+        MockHttpServletResponse response = new MockHttpServletResponse();
+        MockFilterChain chain = new MockFilterChain();
+
+        jwtAuthenticationFilter.doFilterInternal(request, response, chain);
+        assertThat(response.getStatus()).isEqualTo(401);
+        String json = JsonPath.read(response.getContentAsString(), "message");
+        assertThat(json).startsWith("The Token has expired ");
+    }
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-rest-without-microservices/src/test/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilterTest.java:86]: also check the status part of the message returned

File: core/java/iesi-rest-without-microservices/src/test/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilterTest.java [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+        MockHttpServletRequest request = new MockHttpServletRequest("GET", "/scripts");
+        request.addHeader(HttpHeaders.AUTHORIZATION, "Bearer " + token);
+        JWTAuthenticationFilter jwtAuthenticationFilter = new JWTAuthenticationFilter(jwtAuthenticationConverter);
+        MockHttpServletResponse response = new MockHttpServletResponse();
+        MockFilterChain chain = new MockFilterChain();
+
+        jwtAuthenticationFilter.doFilterInternal(request, response, chain);
+        assertThat(response.getStatus()).isEqualTo(401);
+        String json = JsonPath.read(response.getContentAsString(), "message");
+        assertThat(json).startsWith("The Claim 'iss' value doesn't match the required issuer.");
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-rest-without-microservices/src/test/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilterTest.java:54]: also check the status part of the message returned

File: core/java/iesi-rest-without-microservices/src/test/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilterTest.java [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+        MockHttpServletRequest request = new MockHttpServletRequest("GET", "/scripts");
+        request.addHeader(HttpHeaders.AUTHORIZATION, "Bearer " + token);
+        JWTAuthenticationFilter jwtAuthenticationFilter = new JWTAuthenticationFilter(jwtAuthenticationConverter);
+        MockHttpServletResponse response = new MockHttpServletResponse();
+        MockFilterChain chain = new MockFilterChain();
+
+        jwtAuthenticationFilter.doFilterInternal(request, response, chain);
+        assertThat(response.getStatus()).isEqualTo(401);
+        String json = JsonPath.read(response.getContentAsString(), "message");
+        assertThat(json).endsWith("doesn't have a valid JSON format.");
```
Comment (BerrevoetsRobbe) [inline core/java/iesi-rest-without-microservices/src/test/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilterTest.java:38]: also check the status part of the message returned

File: core/java/iesi-rest-without-microservices/src/main/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilter.java (hunk 3 above)
Comment (BerrevoetsRobbe) [reply core/java/iesi-rest-without-microservices/src/main/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilter.java:30]: bump

File: core/java/iesi-rest-without-microservices/src/main/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilter.java (hunk 2 above)
Comment (BerrevoetsRobbe) [reply core/java/iesi-rest-without-microservices/src/main/java/io/metadew/iesi/server/rest/configuration/security/jwt/JWTAuthenticationFilter.java:28]: not true. If no Authorization header is present in the request or the Authorization header does not start with Bearer, the authentication object will be null. This behavior is expected for endpoints that are whitelisted from authorization e.g. /actuator/health. That's why we just call the next filter in the chain. Also, should there have been other filters that have already set the authentication object of SecurityContextHolder.getContext(), there value will be overridden by a null value. This is not wanted behavior

**Review by BerrevoetsRobbe — COMMENTED** [review]
approved
